sfc-aqua/quisp
# Setup opp test

This PR sets up opp_test for module tests.
For using opp_test, we need to build quisp as a library, not executable.
And then opp_test compiles test code and links with the library.
Finally, we can run the test program.

## step to run module tests
1. ensure pkg-config can find the `eigen3.pc` file. You can check with `$ pkg-config eigen3 --validate`.
1. run `$ make lib` at the repository root
1. then you can find `quisp/out/clang-debug/quisp_dbg.dylib` or `quisp/out/clang-debug/quisp_dbg.so` (if you use clang compiler)
1. run `make run-module-test`.

## pkg-config for resolving eigen3 library/header path
now we need to set up the OMNet++IDE to pass the eigen3's path like `/usr/include/eigen3` as which is written in [build doc](https://github.com/sfc-aqua/quisp/blob/master/doc/Build_on_docker.md).
But it's troublesome, and the path depends on where a user installs the eigen3 library.
We can solve this problem by introducing [pkg-config](https://www.freedesktop.org/wiki/Software/pkg-config/).

pkg-config knows where the library is by reading the .pc file in`PKG_CONFIG_PATH`.
In the build process, I use pkg-config to set the library/include path.
So now we don't need to set up the eigen3 path in OMNet++IDE.
In docker, pkg-config can find the eigen3.pc file in default.

## the eigen3.pc file location and `PKG_CONFIG_PATH`
In my macOS and HomeBrew installs the eigen3.pc file into `/usr/local/opt/eigen/share/pkgconfig/`.
so I set `export PKG_CONFIG_PATH=/usr/local/opt`
In my ubuntu, pkg-config can find the eigen3.pc file.

## build process and pkg-config
OMNet++ doesn't provide the way to build both library and executable files in one `Makefile`.
This means we need to run `opp_makemake` to change the configuration.
If we use only OMNet++IDE, there's no problem, but it becomes a problem when we need to test on CI because we need to re-generate `Makefile` by `opp_makemake`.
OMNet++IDE stores the include paths into the `.oppbuildspec` file. This is an XML file, so I considered using this file to feed the paths to `opp_makemake`, but it makes the build process complicated and needs to set up OMNet++IDE. It's not suitable for CI.

Fortunately, `opp_makemake` provides [`makefrag`](https://doc.omnetpp.org/omnetpp/manual/#sec:makefrag) us the ability to customize `Makefile`.
Now `quisp/makefrag` file has a Makefile fragment which contains a fragment to retrieve the Eigen path.

## test structure
TODO

see also [OMNet++ doc](https://doc.omnetpp.org/omnetpp/manual/#sec:testing:opptest) and [INET framework's doc](https://inet.omnetpp.org/TestSuite.html)

---
TODO:
- [x] ~~make version check~~
- [x] ~~load options for opp_makemake from `quisp/.oppbuildspec`~~
- [x] remove tests code from build target
- [x] makefile rule to run all opp_test
